### ISSUE TITLE
Use `method<-` to register methods on S3 generics for base classes

### ISSUE THREAD

**Issue by t-kalinowski** [RConsortium/S7#455]
Currently, this raises an error:
``` r
library(S7)
an_s3_generic <- function(x) UseMethod("an_s3_generic")
method(an_s3_generic, S7::class_character) <- function(x) "foo"
#> Error: When registering methods for S3 generic an_s3_generic(), signature must be an S7 class, not a base type.
```

It would be nice if this worked and `method<-` was truly a single interface for registering any kind of method.

**Comment by hadley** [comment]
I can't remember exactly why but I think I had a good reason for doing this.

**Comment by t-kalinowski** [comment]
I think after #422, implementing this shouldn’t be too difficult, and we can quickly rediscover why it's not allowed 😆.

**Comment by lawremi** [comment]
Agreed that this would need to happen if we really see S7 replacing the legacy S3 registration mechanism.

**Comment by hadley** [comment]
Analysis from claude:

| Signature | S7 generic | S3 generic (`UseMethod`) | S3 Ops/internal (`+`, `[`, …) | S4 generic (`setGeneric`) |
|---|---|---|---|---|
| S7 class (`new_class("Foo")`) | ✓ | ✓ | ✓ | ✓ (S4 reg) |
| Base type (`class_character`, …) | ✓ | ✗ | ✓ | #642 |
| S3 class (`new_S3_class("foo")`) | ✓ | ✗ | ✓ | ✓ (S4 reg) |
| S7 union (`class_numeric`, `class_atomic`, …) | ✓ (flattened) | #636 | ✓ | ✗ |
| S4 class object | ✓ | ✗ | ✓ | ✓ |
| `class_any` | ✓ | ✗ | ✓ | ✗ |
| `class_missing` | ✓ | ✗ (only needed for multi-dispatch) | ✓ | ✗ |
| `NULL` | ✓ | ✗ | ✓ | ✗ |
| List signature (multi-dispatch) | ✓ (if generic has ≥2 dispatch args) | ✗ (S3 is single-dispatch) | ✓ (Ops dispatches on `e1, e2`) | ✓ |

**Notes:**

- **S3 Ops/internal column:** these *look* like S3 generics but `as_S3_generic()` rewrites them into entries of `base_ops` / `base_matrix_ops`, which are real S7 generics. So they accept everything the S7 column accepts.
- **External generics** (`new_external_generic()`): resolve to the underlying generic via `getFromNamespace()` and inherit that column's behavior. For deferred (`.onLoad`) registration the same gates fire when the upstream package loads, so the failure mode is "package installs, then `.onLoad` blows up later."
- **S4 + S3 wrapper / S3 class:** the "(S4 reg)" requirement is currently a manual step; could plausibly be auto-registered.

---

I'll chip away at this in a series of PRs.

**Comment by hadley** [comment]
After the two PRs in flight are merged, I think I can close this PR by doing a little refactoring, e.g. creating separate S3 and S4 classes, and doing the flattening earlier.
